File: agent/packages/schema/src/yidao.ts
import { Type, type Static } from "@sinclair/typebox";

import { MeridianId } from "./cultivation.js";
import { validate, type ValidationResult } from "./validate.js";

export const YidaoSkillIdV1 = Type.Union([
  Type.Literal("meridian_repair"),
  Type.Literal("contam_purge"),
  Type.Literal("emergency_resuscitate"),
  Type.Literal("life_extension"),
  Type.Literal("mass_meridian_repair"),
]);
export type YidaoSkillIdV1 = Static<typeof YidaoSkillIdV1>;

export const YidaoEventKindV1 = Type.Union([
  Type.Literal("meridian_heal"),
  Type.Literal("contam_purge"),
  Type.Literal("emergency_resuscitate"),
  Type.Literal("life_extension"),
  Type.Literal("mass_heal"),
  Type.Literal("karma_accumulation"),
  Type.Literal("medical_contract"),
]);
export type YidaoEventKindV1 = Static<typeof YidaoEventKindV1>;

export const MedicalContractStateV1 = Type.Union([
  Type.Literal("stranger"),
  Type.Literal("patient"),
  Type.Literal("long_term_patient"),
  Type.Literal("bonded"),
]);
export type MedicalContractStateV1 = Static<typeof MedicalContractStateV1>;

export const YidaoEventV1 = Type.Object(
  {
    v: Type.Literal(1),
    kind: YidaoEventKindV1,
    tick: Type.Integer({ minimum: 0 }),
    medic_id: Type.String({ minLength: 1, maxLength: 128 }),
    patient_ids: Type.Array(Type.String({ minLength: 1, maxLength: 128 })),
    skill: YidaoSkillIdV1,
    meridian_id: Type.Optional(MeridianId),
    success_count: Type.Integer({ minimum: 0 }),
    failure_count: Type.Integer({ minimum: 0 }),
    qi_transferred: Type.Number({ minimum: 0 }),
    contam_reduced: Type.Number({ minimum: 0 }),
    hp_restored: Type.Number({ minimum: 0 }),
    karma_delta: Type.Number({ minimum: 0 }),
    medic_qi_max_delta: Type.Number({ maximum: 0 }),
    patient_qi_max_delta: Type.Number({ maximum: 0 }),
    contract_state: Type.Optional(MedicalContractStateV1),
    detail: Type.String({ minLength: 1, maxLength: 256 }),
  },
  { additionalProperties: false },
);
export type YidaoEventV1 = Static<typeof YidaoEventV1>;

export const HealerNpcAiStateV1 = Type.Object(
  {
    healer_id: Type.String({ minLength: 1, maxLength: 128 }),
    active_action: Type.String({ minLength: 1, maxLength: 64 }),
    queue_len: Type.Integer({ minimum: 0 }),
    reputation: Type.Integer(),
    retreating: Type.Boolean(),
  },
  { additionalProperties: false },
);
export type HealerNpcAiStateV1 = Static<typeof HealerNpcAiStateV1>;

export const YidaoHudStateV1 = Type.Object(
  {
    healer_id: Type.String({ minLength: 1, maxLength: 128 }),
    reputation: Type.Integer(),
    peace_mastery: Type.Number({ minimum: 0, maximum: 100 }),
    karma: Type.Number({ minimum: 0 }),
    active_skill: Type.Union([YidaoSkillIdV1, Type.Null()]),
    patient_ids: Type.Array(Type.String({ minLength: 1, maxLength: 128 })),
    patient_hp_percent: Type.Union([
      Type.Number({ minimum: 0, maximum: 1 }),
      Type.Null(),
    ]),
    patient_contam_total: Type.Union([Type.Number({ minimum: 0 }), Type.Null()]),
    severed_meridian_count: Type.Integer({ minimum: 0 }),
    contract_count: Type.Integer({ minimum: 0 }),
    mass_preview_count: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type YidaoHudStateV1 = Static<typeof YidaoHudStateV1>;

export function validateYidaoEventV1Contract(data: unknown): ValidationResult {
  return validate(YidaoEventV1, data);
}

export function validateHealerNpcAiStateV1Contract(data: unknown): ValidationResult {
  return validate(HealerNpcAiStateV1, data);
}

export function validateYidaoHudStateV1Contract(data: unknown): ValidationResult {
  return validate(YidaoHudStateV1, data);
}

File: server/src/schema/yidao.rs
//! 医道 v1 跨栈契约。

use serde::{Deserialize, Serialize};

use crate::cultivation::components::MeridianId;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum YidaoSkillIdV1 {
    MeridianRepair,
    ContamPurge,
    EmergencyResuscitate,
    LifeExtension,
    MassMeridianRepair,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum YidaoEventKindV1 {
    MeridianHeal,
    ContamPurge,
    EmergencyResuscitate,
    LifeExtension,
    MassHeal,
    KarmaAccumulation,
    MedicalContract,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MedicalContractStateV1 {
    Stranger,
    Patient,
    LongTermPatient,
    Bonded,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct YidaoEventV1 {
    pub v: u8,
    pub kind: YidaoEventKindV1,
    pub tick: u64,
    pub medic_id: String,
    pub patient_ids: Vec<String>,
    pub skill: YidaoSkillIdV1,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meridian_id: Option<MeridianId>,
    pub success_count: u32,
    pub failure_count: u32,
    pub qi_transferred: f64,
    pub contam_reduced: f64,
    pub hp_restored: f32,
    pub karma_delta: f64,
    pub medic_qi_max_delta: f64,
    pub patient_qi_max_delta: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub contract_state: Option<MedicalContractStateV1>,
    pub detail: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct HealerNpcAiStateV1 {
    pub healer_id: String,
    pub active_action: String,
    pub queue_len: u32,
    pub reputation: i32,
    pub retreating: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct YidaoHudStateV1 {
    pub healer_id: String,
    pub reputation: i32,
    pub peace_mastery: f32,
    pub karma: f64,
    pub active_skill: Option<YidaoSkillIdV1>,
    pub patient_ids: Vec<String>,
    pub patient_hp_percent: Option<f32>,
    pub patient_contam_total: Option<f64>,
    pub severed_meridian_count: u32,
    pub contract_count: u32,
    pub mass_preview_count: u32,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn yidao_event_roundtrips_with_no_unknown_fields() {
        let event = YidaoEventV1 {
            v: 1,
            kind: YidaoEventKindV1::MassHeal,
            tick: 42,
            medic_id: "offline:Healer".to_string(),
            patient_ids: vec![
                "offline:PatientA".to_string(),
                "offline:PatientB".to_string(),
            ],
            skill: YidaoSkillIdV1::MassMeridianRepair,
            meridian_id: Some(MeridianId::Lung),
            success_count: 2,
            failure_count: 0,
            qi_transferred: 120.0,
            contam_reduced: 0.0,
            hp_restored: 0.0,
            karma_delta: 0.2,
            medic_qi_max_delta: -0.04,
            patient_qi_max_delta: 0.0,
            contract_state: None,
            detail: "mass repair".to_string(),
        };
        let text = serde_json::to_string(&event).expect("serialize");
        assert!(text.contains("\"mass_meridian_repair\""));
        let back: YidaoEventV1 = serde_json::from_str(&text).expect("deserialize");
        assert_eq!(back.kind, YidaoEventKindV1::MassHeal);
        assert_eq!(back.patient_ids.len(), 2);
    }

    #[test]
    fn yidao_hud_state_serializes_active_skill() {
        let state = YidaoHudStateV1 {
            healer_id: "npc:doctor".to_string(),
            reputation: 12,
            peace_mastery: 48.0,
            karma: 3.5,
            active_skill: Some(YidaoSkillIdV1::LifeExtension),
            patient_ids: vec!["offline:NearDeath".to_string()],
            patient_hp_percent: Some(0.5),
            patient_contam_total: Some(1.25),
            severed_meridian_count: 1,
            contract_count: 2,
            mass_preview_count: 0,
        };
        let value = serde_json::to_value(state).expect("serialize");
        assert_eq!(value["active_skill"], "life_extension");
    }
}
